opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py

```diff
@@ -164,60 +169,128 @@ def on_end(self, span: Span) -> None:
 
     def worker(self):
         timeout = self.schedule_delay_millis / 1e3
+        flush_request = None  # type: typing.Optional[_FlushRequest]
         while not self.done:
-            if (
-                len(self.queue) < self.max_export_batch_size
-                and not self._flushing
-            ):
-                with self.condition:
+            with self.condition:
```

Comment:
What is the purpose of having `with self.condition` encapsulating the whole code segment, instead of putting it right  before `self.condition.wait`?

Reply:
acquiring the `flush_request` needs to be synchronized. in case of a `flush_request` we want to avoid `self.condition.wait` so we need to have the statements for acquiring the `flush_request` (call to `_get_and_unset_flush_request`) and the wait call in the same `with self.condition` block to be sure that no flush request was made in between those statements. we also need to re-check if a `flush_request` was made while waiting. so this needs also go inside the same `with self.condition` block. 

i guess to minimize the code within `with self.condition` something like the following can be done
```python
        while not self.done:
            with self.condition:
                if self.done:
                    # done flag may have changed, avoid waiting
                    break
                flush_request = self._get_and_unset_flush_request()
            if (
                len(self.queue) < self.max_export_batch_size
                and flush_request is None
            ):
                with self.condition:
                    if self.done:
                        # done flag may have changed, avoid waiting
                        break
                    flush_request = self._get_and_unset_flush_request()
                    if not flush_request:
                        self.condition.wait(timeout)
                        flush_request = self._get_and_unset_flush_request()
                if not self.queue:
                    # spurious notification, let's wait again
                    self._notify_flush_request_finished(flush_request)
                    flush_request = None
                    continue
                if self.done:
                    # missing spans will be sent when calling flush
                    break

            # subtract the duration of this export call to the next timeout
            start = time_ns()
            ...
```
but this one looks just more bloated and i'm not sure if there are any noticable performance gains.